contracts/beacon/Voucher.sol

```diff
@@ -164,11 +184,69 @@ contract Voucher is OwnableUpgradeable, IVoucher {
         return dealId;
     }
 
+    function claim(bytes32 dealId, uint256 index) external {
+        VoucherStorage storage $ = _getVoucherStorage();
+        IVoucherHub voucherHub = IVoucherHub($._voucherHub);
+        address iexecPoco = voucherHub.getIexecPoco();
+        // Check if task is already claimed.
+        bytes32 taskId = keccak256(abi.encodePacked(dealId, index));
+        require(!$._claimedTasks[taskId], "Voucher: task already claimed");
+        // Get deal details from PoCo.
+        (bool isBoostDeal, address requester, uint256 volume, uint256 dealPrice) = _getDealDetails(
+            iexecPoco,
+            dealId
+        );
+        require(requester != address(0), "Voucher: deal not found");
+        // Check wether the deal was matched by the voucher or not.
+        VoucherMatchedDeal memory voucherMatchedDeal = $._voucherMatchedDeals[dealId];
+        if (voucherMatchedDeal.exists) {
+            // The deal was matched by the voucher.
+            // It is either fully, partially, or not sponsored.
+            IexecLibCore_v5.Task memory task = IexecPocoAccessors(iexecPoco).viewTask(taskId);
+            if (task.status != IexecLibCore_v5.TaskStatusEnum.FAILED) {
+                // Claim task on PoCo because it's not already claimed.
+                // Tx will be reverted by the PoCo if the claim is not valid.
+                _claimTaskOnPoco(iexecPoco, isBoostDeal, dealId, index);
+            }
+            if (dealPrice != 0) {
+                // For partially sponsored or non sponsored deals, send the requester's contribution
+                // back to their iExec account because all funds are, by default, sent to the sponsor
+                // (the voucher in this case).
+                //
+                // The division yields no remainder since dealPrice = taskPrice * volume.
+                uint256 taskPrice = dealPrice / volume;
+                // A positive remainder is possible when the voucher balance is less than
+                // the sponsorable amount. Min(balance, dealSponsoredAmount) is computed
+                // at match orders.
+                // TODO !! do something with the remainder.
+                uint256 taskSponsoredAmount = voucherMatchedDeal.sponsoredAmount / volume;
+                if (taskSponsoredAmount != 0) {
+                    // If the voucher did fully/partially sponsor the deal then mint voucher
+                    // credits back.
+                    voucherHub.refundVoucher(taskSponsoredAmount);
+                }
+                if (taskSponsoredAmount < taskPrice) {
+                    // If the deal was not sponsored or partially sponsored
+                    // by the voucher then send the non-sponsored part back
+                    // to the requester.
+                    IERC20(iexecPoco).transfer(requester, taskPrice - taskSponsoredAmount);
+                }
+            }
+            $._claimedTasks[taskId] = true;
+        } else {
```

Comment:
Do we want people to be able to claim "any" other than theirs own task ?

Reply:
The same thing as in PoCo yes.
If the requester is not sure if the voucher has sponsored it's task, it can send the claim to the voucher which in turn claims it for them even if it wasn't the sponsor. More of a UX simplification. Bu I'm open to discussion.